### PR TITLE
Validate checkout fields before creating a payment method

### LINKED ISSUE #8192: Shortcode checkout displays Stripe error when payment information is entered but billing fields are empty
### Describe the bug

When in the shortcode checkout, if payment information is entered but required billing fields are empty, [the call to `createPaymentMethod`](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/classic/payment-processing.js#L178) will throw a user-unfriendly error that will be displayed to the user. This occurs because [we instruct the payment element not to collect any billing fields](https://github.com/Automattic/woocommerce-payments/blob/develop/client/checkout/utils/upe.js#L84-L118), since we will collect this information from the checkout fields instead. If these fields are missing when we create the payment method, Stripe's API will throw an error that we will display immediately to the user. 

Any payment method, name missing
_Place order with any payment method, but name fields are empty_

Afterpay, email missing
_Place order with Afterpay with name present, but email address field is empty_

Affirm, address missing
_Place order with Affirm with name present, but address field is empty_

Instead of showing this incomprehensible error to the shopper, we should display the usual WooCommerce form field validation errors instead.

Usual WC validation errors
_Expected WC validation errors_

#### First error message (any PM)

1. Add an item to cart, proceed to shortcode checkout.
2. Ensure all billing fields are empty, select any payment method, and fill out payment method fields, if any are present/required.
3. Select Place order.
4. Notice error stating, `You specified "never" for fields.billing_details.name when creating the payment Element, but did not pass params.payment_method_data.billing_details.name when calling stripe.createPaymentMethod(). If you opt out of collecting data via the payment Element using the fields option, the data must be passed in when calling stripe.createPaymentMethod().`

#### More error messages (Affirm/Afterpay)

1. Ensure Affirm & Afterpay are enabled at checkout and enable a currency that is compatible with both payment methods (e.g. USD).
2. Add an item or items to cart that satisfy amount limits for both payment methods (e.g. total amount greater than $50, less than $4,000).
3. Select either Affirm or Afterpay, leaving all billing fields empty.
4. Select Place order.
5. Notice first error mentioned above.
6. Select Afterpay, fill name fields, but leave email and address fields empty.
7. Select Place order.
8. Notice error message stating, `Missing required param: billing_details[email].`
9. Select Affirm, with name fields filled, and address fields empty.
10. Select Place order.
11. Notice error message stating, `Missing required param: billing_details[address][line1].`

### Actual behavior

An ugly and confusing error message is displayed to shoppers when billing fields are empty.

### Screenshots

See above.

### Expected behavior

Expected and formatted WooCommerce validation errors should be shown to shoppers instead.

### Desktop (please complete the following information):

* OS: [e.g. iOS] Any
* Browser [e.g. chrome, safari]: Any
* Version [e.g. 22]: Latest

### Smartphone (please complete the following information):

* Device: [e.g. iPhone6] Any
* OS: [e.g. iOS8.1] Any
* Browser [e.g. stock browser, safari]: Any
* Version [e.g. 22]: Latest

### Additional context

CC @pierorocca

### PR DESCRIPTION
**Pull request by mdmoore**
Fixes #8192 

#### Changes proposed in this Pull Request

This adds validation to the checkout data that's used to create a payment method on checkout. Currently, empty fields are passed in and Stripe error is displayed on checkout, rather than the WC core checkout validation messages. With this validation in place, an error will be thrown and the form will be submitted to allow core to validate and display it's own error messages.

#### Testing instructions

**Correct Error Messages**
Screenshot 2024-02-22 at 3 00 10 PM

1. Add an item to cart, proceed to shortcode checkout.
2. Ensure all billing fields are empty, select any payment method(new card is fine), and fill out payment method fields, if any are present/required.
3. Select Place order.
4. Notice the standard `<field> is a required field` message and not something like this:

`You specified "never" for fields.billing_details.name when creating the payment Element, but did not pass params.payment_method_data.billing_details.name when calling stripe.createPaymentMethod(). If you opt out of collecting data via the payment Element using the fields option, the data must be passed in when calling stripe.createPaymentMethod().`

5. Repeat the same with AfterPay and Affirm, but leave email and address fields empty
6. Verify that the core error messages are displayed and not anything like:

`Missing required param: billing_details[email].` 
or
`Missing required param: billing_details[address][line1].`

-------------------

- [x] Run `npm run changelog` to add a changelog file, choose `patch` to leave it empty if the change is not significant. You can add multiple changelog files in one PR by running this command a few times. 
- [ ] Covered with tests (or have a good reason not to test in description ☝️)
- [ ] Tested on mobile (or does not apply)

**Post merge**

- [ ] Link to testing instructions from [release testing doc](https://github.com/Automattic/woocommerce-payments/wiki/Release-testing-instructions) following [these instructions](https://github.com/Automattic/woocommerce-payments/wiki/How-to-write-good-manual-testing-scenarios) : _Add link here / 'QA Testing Not Applicable'_
- [ ] Add or update [critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Critical-flows) and [testing instructions for critical flows](https://github.com/Automattic/woocommerce-payments/wiki/Testing-instructions-for-critical-flows), if applicable.
- [ ] Add what's changed (description, screenshot, demo videos etc.) to the release announcement post, if applicable.

### CODE REVIEW

**Comment by botwoo** [comment]
### Test the build 

#### Option 1. Jetpack Beta

- Install and activate [Jetpack Beta](https://jetpack.com/download-jetpack-beta/).
- Use this build by searching for PR number `8258` or branch name `fix/8192-payment-element-error-handling` in your-test.site/wp-admin/admin.php?page=jetpack-beta&plugin=woocommerce-payments          

#### Option 2. Jurassic Ninja - available for logged-in A12s

:rocket: [Launch a JN site with this branch](https://jurassic.ninja/create/?jetpack-beta&shortlived&nojetpack&woocommerce&woocommerce-payments-dev-tools&branches.woocommerce-payments=fix/8192-payment-element-error-handling) :rocket:

:information_source: Install this [Tampermonkey script](https://github.com/Automattic/woocommerce-payments/tree/develop/bin/wcpay-live-branches) to get more options.

---

Build info: 

- Latest commit: 6b7351bfdb386a88f8288fd332d55ce2999a600b
- Build time: 2024-02-29 11:55:50 UTC

_Note: the build is updated when a new commit is pushed to this PR._

**Comment by cs9k** [comment]
Seems like there is an issue regarding how merchant requests are handled, could you look into it? This would fix my issue #8270 as well

**Comment by mdmoore** [comment]
@frosso Feel free to push any changes here if you want and we can review in parallel to get this merged ASAP.

File: client/checkout/classic/payment-processing.js [hunk 1]
```diff
@@ -93,6 +93,36 @@ export function validateElements( elements ) {
...
+			throw new Error( 'Missing required field' );
+		}
+	}
+
+	const requiredAddressFields = [
+		'city',
+		'country',
+		'line1',
+		'postal_code',
+		'state',
```
Comment (frosso) [inline client/checkout/classic/payment-processing.js]: In some countries, the `state` field is optional (e.g.: Germany)

Could you please confirm if you're seeing this message displayed at checkout, even when filling in all the required details?
Screenshot 2024-02-27 at 6 22 15 PM

**Comment by frosso** [comment]
@mdmoore I like your idea of validating the missing fields. What do you think of leveraging the `getUPEConfig( 'enabledBillingFields' )` to get the enabled fields, and doing the validation on this function, instead? https://github.com/Automattic/woocommerce-payments/blob/05ddc07ed6a983cb8117bfc40edcd3e5a60fc3c2/client/checkout/classic/event-handlers.js#L73-L75

I am thinking that by leveraging the `enabledBillingFields`, we would use the existing configuration, making it aligned with the logic in `createStripePaymentMethod()`.
This way, we also don't need to check on the error message in the body of `processPayment()`, from an exception being thrown somewhere else in the codebase.

To get an idea, I implemented these changes here: https://github.com/Automattic/woocommerce-payments/pull/8281

What do you think? I'm leaving it in draft, but we can discuss further if those suggestions are ok!

File: client/checkout/classic/payment-processing.js (hunk 1 above)
Comment (mdmoore) [reply client/checkout/classic/payment-processing.js]: Yep. This makes sense. Thanks for the example.

**Comment by mdmoore** [comment]
@frosso

>To get an idea, I implemented these changes here: https://github.com/Automattic/woocommerce-payments/pull/8281
>What do you think? I'm leaving it in draft, but we can discuss further if those suggestions are ok!

Yes, it's a great idea and much cleaner solution. I've tested it out and it's working perfectly. I'm in favor of either merging #8281 here or directly to `develop`. Which ever you prefer.

**Comment by frosso** [comment]
@mdmoore I went ahead and merged my changes in this PR. I couldn't change my PR's base branch, github is having some issues right now.
I waited to ensure the e2e were now passing.
There is a failure on the tests, which I'm fixing on https://github.com/Automattic/woocommerce-payments/pull/8294 and is not related to these changes.

I assigned someone else to review, so that we could have a fair merge :) Thanks!

**Review by timur27 — APPROVED** [review]
Thanks both for the contributions! 🚀 All tests well, Blocks are unrelated (an obvious thing but leaving it here in case I end up being wrong and someone could correct me). 

Code looks good and is so readable. One suggestion is to move the function _from_ the script responsible for events handling _to_ `utils/upe.js` where currently all similar helper functions belong like e.g. `isPaymentMethodRestrictedToLocation()` or `isUsingSavedPaymentMethod()`. Moving it to `utils/upe.js` will also allow to test it quite easily. 

Other than that the PR looks good and I'm approving it while keeping the above suggestions to be applied/rejected depending on how you all view things.

**Comment by frosso** [comment]
Thanks @timur27 ! I like your suggestion of moving the function down, it makes sense, so I went ahead and moved it 👍

**Comment by mdmoore** [comment]
Thanks @frosso for your solution on this and @timur27 for reviewing and collaborating. 🏆s to you both!

**Pull request merged**